### PR TITLE
Recycler view mail list

### PR DESCRIPTION
**Pull request by kamil0074**
In response on issue:
https://github.com/k9mail/k-9/issues/2807

Changed ListView to RecycleView. Split MessageListFragment to more classes. Wrapped existing CursorAdapter i  RecycleViewAdapter. Added delete by swipe functionality and tested behaviour.

### CODE REVIEW

File: k9mail/build.gradle [hunk 1]
```diff
@@ -28,7 +32,8 @@ dependencies {
     compile "com.squareup.okio:okio:${okioVersion}"
     compile 'commons-io:commons-io:2.4'
     compile "com.android.support:support-v4:${androidSupportLibraryVersion}"
-    compile 'org.jsoup:jsoup:1.10.2'
+	compile "com.android.support:recyclerview-v7:25.2.0"
```
Comment (philipwhiuk) [inline k9mail/build.gradle]: Indentation.

File: build.gradle [hunk 2]
```diff
@@ -1,7 +1,11 @@
 buildscript {
     repositories {
         jcenter()
-    }
+		maven{
```
Comment (philipwhiuk) [inline build.gradle]: Indentation

Comment (philipwhiuk) [reply build.gradle]: Also, do we actually need it in the main build.gradle?

File: k9mail/build.gradle [hunk 3]
```diff
@@ -38,6 +43,7 @@ dependencies {
     compile "com.jakewharton.timber:timber:${timberVersion}"
     compile 'net.jcip:jcip-annotations:1.0'
 
+
```
Comment (philipwhiuk) [inline k9mail/build.gradle]: Uneccessary extra line

File: k9mail/src/main/java/com/fsck/k9/fragment/ICursorContextMenuSupplier.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package com.fsck.k9.fragment;
+import android.database.Cursor;
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/fragment/ICursorContextMenuSupplier.java]: Missing new line.

File: k9mail/src/main/java/com/fsck/k9/fragment/ICursorContextMenuSupplier.java [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+import android.view.ContextMenu;
+/**
+ Created by Kamil Rajtar on 05.11.17. */
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/fragment/ICursorContextMenuSupplier.java]: Remove this and similar such docs - we attribute via changelogs and Git commits, not in the code.

File: k9mail/src/main/java/com/fsck/k9/fragment/ICursorContextMenuSupplier.java [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+
+public interface ICursorContextMenuSupplier{
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/fragment/ICursorContextMenuSupplier.java]: lint rule - missing space after class name. Also present in other files

File: k9mail/src/main/java/com/fsck/k9/fragment/MessageListHandler.java [hunk 7]
```diff
@@ -129,8 +128,10 @@ public void handleMessage(android.os.Message msg) {
                 break;
             }
             case ACTION_RESTORE_LIST_POSITION: {
-                fragment.listView.onRestoreInstanceState((Parcelable) msg.obj);
-                break;
+				throw new UnsupportedOperationException("Not restored yet");
+				//TODO restore
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/fragment/MessageListHandler.java]: TODO

File: k9mail/src/main/java/com/fsck/k9/fragment/MessageRecycleViewAdapter.java [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+	private final Context context;
+	private final CursorAdapter internalAdapter;
+	private final IOnMessageClickListener onMessageClickListener;
+	private final ICursorContextMenuSupplier cursorContextMenuSupplier;
+	private int selectedItem=-1;
+	private FooterViewHolder footerViewHolder;
+
+	public MessageRecycleViewAdapter(final Context context,
+									 final MessageListFragment messageListFragment,
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/fragment/MessageRecycleViewAdapter.java]: Indentation.

File: k9mail/src/main/java/com/fsck/k9/fragment/MessageRecycleViewAdapter.java [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+public class MessageRecycleViewAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder>{
+	private static final int TYPE_MESSAGE_ITEM=0;
+	private static final int TYPE_FOOTER_ITEM=1;
+
+	private final Context context;
+	private final CursorAdapter internalAdapter;
+	private final IOnMessageClickListener onMessageClickListener;
+	private final ICursorContextMenuSupplier cursorContextMenuSupplier;
+	private int selectedItem=-1;
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/fragment/MessageRecycleViewAdapter.java]: code formatting. Also present elsewhere - space before and after operators.

File: k9mail/src/main/java/com/fsck/k9/fragment/MessageRecycleViewAdapter.java [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+			}
+		});
+	}
+
+	private void tryMoveSelection(final RecyclerView recyclerView,final int direction){
+		final int nextSelectItem=selectedItem+direction;
+
+		if(nextSelectItem>=0&&nextSelectItem<getItemCount()){
+			notifyItemChanged(selectedItem);
+//			recyclerView.findViewHolderForAdapterPosition(selectedItem).itemView.setBackgroundColor(
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/fragment/MessageRecycleViewAdapter.java]: Commented out code. Remove if unused.

File: k9mail/src/main/java/com/fsck/k9/fragment/MessageRecycleViewAdapter.java [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+	public boolean isEmpty(){
+		return internalAdapter.isEmpty();
+	}
+
+	public void addFooterView(final View footerView){
+		footerViewHolder=new FooterViewHolder(footerView);
+		notifyDataSetChanged();
+	}
+
+	public class MessageViewHolder extends RecyclerView.ViewHolder{
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/fragment/MessageRecycleViewAdapter.java]: Are these actually valuable if all they do is inherit their super classes?

File: k9mail/src/main/res/values/colors.xml [hunk 12]
```diff
@@ -6,5 +6,6 @@
     <color name="message_list_widget_divider">#e5e5e5</color>
     <color name="message_list_widget_text_read">#444444</color>
     <color name="message_list_widget_text_unread">#000000</color>
+    <color name="swipe_delete">#d50000</color>
```
Comment (philipwhiuk) [inline k9mail/src/main/res/values/colors.xml]: Have you tested this in dark theme?

**Comment by qrovira** [comment]
I cleaned up part of the work on a separate branch at [qrovira/RecyclerViewMailList](https://github.com/qrovira/k-9/tree/RecyclerViewMailList), mostly so I could squash it to read the diff. I also tried to leave out any cleanups/improvements unrelated to eg. formatting/final/etc. Some of those might be well worth bringing back in a separate PR.

There's still a few TODOs left around as well, so this still needs work before it's tidy to merge.

The reason I was looking into this is because I started taking a look at #1426, and implemented that on the current view as an option (for both swipe left and right). Would it make sense to build that on top of this PR instead?

File: k9mail/src/main/res/values/colors.xml (hunk 12 above)
Comment (kamil0074) [reply k9mail/src/main/res/values/colors.xml]: Yes and it is Ok

**Comment by kamil0074** [comment]
I cleaned the code and it is prepared to merge

**Comment by cketti** [comment]
Thanks a lot for this. We do want to replace `ListView` with `RecyclerView`.

But this pull request contains too many changes which makes it hard to review properly. It should be split into at least three different pull requests:
- Clean up existing code
- Replace `ListView` with `RecyclerView`
- Add 'swipe to delete' functionality. This conflicts with the 'swipe to select' functionality we currently have. I'm not yet sure what's the best way to introduce this feature without surprising users of the existing feature.

A couple of notes:
- We don't use interfaces prefixed with `I`.
- We don't use `final` on method parameters and local variables.
- The smaller your pull request the higher the chance of getting fast feedback and getting the code merged.

I'm closing this pull request. Please do create a new one which just cleans up the code or just does the transition to `RecyclerView` without cleaning up existing code. And please only create the next PR once the previous one has been reviewed and merged.

**Pull request closed without merging**